### PR TITLE
Ct 2423 generate acknowledgement letter text

### PR DESCRIPTION
**Pull request by Polsonby**
## Description
Letter templates and letter generation - new model for letter templates. In the end decided to use ERB since it's built in to Ruby hence no new dependencies in the app. 

The letter_template takes a values object, which can either be a case or specially constructed - we'll be adding a Letter class which wraps a case in order to add features subsequently such as knowing when a previous letter was sent and referring to it in a template, but for now it takes the case
 
## Self-review checklist

* [x] (1) Quick stakeholder demo done OR
* [ ] (2) ...bug with before and after screenshots
* [ ] (3) Tests passing
* [x] (4) Branch ready to be merged (not work in progress)
* [x] (5) No superfluous changes in diff
* [x] (6) No TODO's without new ticket numbers

### CODE REVIEW

File: app/models/letter_template.rb [hunk 1]
```diff
@@ -0,0 +1,13 @@
+class LetterTemplate < ApplicationRecord
+  validates :name, :body, :template_type, presence: true
+
+  enum template_type: {
+    dispatch: 'dispatch',
```
Comment (mseedat-moj) [inline app/models/letter_template.rb]: `despatch` !!

File: config/routes.rb [hunk 2]
```diff
@@ -157,6 +157,11 @@
     end
 
     resources :data_requests
+
+    resource :letters, only: [], path: "letters/:type"  do
+        get "/new", to: 'letters#new'
+        get "/render", to: 'letters#render_letter', as: 'render'
```
Comment (mseedat-moj) [inline config/routes.rb]: Should `render_letter` be `show` ?

File: app/controllers/cases/letters_controller.rb [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+  def render_letter
+    @type = params[:type]
+    unless params[:letter_form] and params[:letter_form][:letter_template_id]
+      flash[:notice] = 'Please select a template.'
+      redirect_to new_case_letters_path(@case.id, @type) and return
+    end
+    letter_template_id = params[:letter_form][:letter_template_id]
+    @letter = Letter.new(letter_template_id)
+    @letter_template = LetterTemplate.find(letter_template_id)
+  end
```
Comment (mseedat-moj) [inline app/controllers/cases/letters_controller.rb]: Would prefer all this logic to either be in the `letter` model or in an appropriate service.

Don't use the `and` operator, use `&&` instead, unless you really do mean `and` ?

Should use strong parameters for consistency with codebase

File: app/models/letter_template.rb (hunk 1 above)
Comment (Polsonby) [reply app/models/letter_template.rb]: Nope :) - they're both valid UK English but Alex and the team chose to settle on `dispatch` e.g.       `ready_to_dispatch`. We still use `de-` in places accidentally, but it's correct here

image

File: config/routes.rb (hunk 2 above)
Comment (Polsonby) [reply config/routes.rb]: Excellent wheeze, never occurred to me. I knew it wasn't create...

File: app/controllers/cases/letters_controller.rb [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+  def new
+    @type = type_name(params[:type])
+    @letter = Letter.new(params[:letter_template_id])
+    @letter_templates = LetterTemplate.where(template_type: @type)
+  end
+
+  def render_letter
+    @type = params[:type]
+    unless params[:letter_form] and params[:letter_form][:letter_template_id]
```
Comment (Polsonby) [inline app/controllers/cases/letters_controller.rb]: Yes, this `and` is used for flow control and short-circuit exits early, never evaluating the right hand side if the left is false - here it prevents the second part from blowing up if params[:letter_form] is nil

File: app/controllers/cases/letters_controller.rb (hunk 3 above)
Comment (Polsonby) [reply app/controllers/cases/letters_controller.rb]: > Would prefer all this logic to either be in the letter model or in an appropriate service.
Will take a look at that, thanks...

> Should use strong parameters for consistency with codebase
This is what seemed strange to me - I went to add params but there are actually lots of places we just access the params directly. So this is IMO consistent with the codebase, which uses both approaches. I've ensured the params are not reflected back to the user in the page, and they're all used safely IMO. But I'll have another look at it, as I agree it's best-practice.

File: app/models/letter_template.rb (hunk 1 above)
Comment (mseedat-moj) [reply app/models/letter_template.rb]: Booo. 51st state here we come

File: app/controllers/cases/letters_controller.rb (hunk 4 above)
Comment (mseedat-moj) [reply app/controllers/cases/letters_controller.rb]: The right hand side will not be executed if `params[:letter_form]` is `true` when using `&&`
You can probably replace
```
unless params[:letter_form] and params[:letter_form][:letter_template_id]
```
with
```
unless params.dig(:letter_form, :letter_template_id)
 ...
end
```

File: app/controllers/cases/letters_controller.rb [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+  private
+
+  def set_case
+    @case = Case::Base.find(params[:case_id])&.decorate
+  end
+
+  def type_name(type)
+    LetterTemplate.template_types[type] || 'unknown'
+  end
+end
```
Comment (mseedat-moj) [inline app/controllers/cases/letters_controller.rb]: `def type_name` should be in the `LetterTemplate` model

File: app/controllers/cases/letters_controller.rb (hunk 4 above)
Comment (Polsonby) [reply app/controllers/cases/letters_controller.rb]: oh nice

**Review by Polsonby — COMMENTED** [review]
Renamed render letter to show and small tweaks. Agree about moving things into service, but would prefer to do this in a follow up PR - the Word doc one is up next(ish)

**Comment by mseedat-moj** [comment]
Generally in the right direction

Some comments in no particular order:

- You seem to have upgraded to Postgress 11? As a result the structure.sql file makes significant changes. Production is on 9.5
- The parameter `letter_template_id` should just be `template_id` so that in use it reads easier (`@letter.template_id`)
- In routes.rb file, the route for `/letters` can be rejigged so that individual `get` paths don't need to be defined
- No Seeds?
- if using ERB templates then does the letter_templates table need to exist? e.g. Easier to design and edit ERB files in the  file system, in the same way ActiveMailer templates are just .erb files in the actual filesystem? Or need to have more meta-data for each template type going forward?
- `render` function currently has an unnecessary temporary variable assignment
```
  def render(values)
    ERB.new(body).result(binding)
  end
```  

- At the moment the routes make a distinction between /dispatch/new and /acknowledgement/new. What's the reason for that?

**Comment by Polsonby** [comment]
Thanks Mo! 

> You seem to have upgraded to Postgress 11? As a result the structure.sql file makes significant changes. Production is on 9.5

I've not upgraded, that's the version that brew installed originally. I noticed this and trying to work up the courage to remove and reinstall the right version! As you'll see from recent commits I've been trying to work through what's different and got there in the end

> The parameter letter_template_id should just be template_id so that in use it reads easier (@letter.template_id)

maybe... I think that's subjective. 

> In routes.rb file, the route for /letters can be rejigged so that individual get paths don't need to be defined

True, I'm not sure it's worth the time though, we got rid of render but there's a word generation coming

> No Seeds?

WIP

> if using ERB templates then does the letter_templates table need to exist? e.g. Easier to design and edit ERB files in the file system, in the same way ActiveMailer templates are just .erb files in the actual filesystem? Or need to have more meta-data for each template type going forward?

Good point, but yes it does need to exist; we have templates of different types and in future might wish to have CMS editing of these. 

> render function currently has an unnecessary temporary variable assignment def render(values)

It's actually necessary

> At the moment the routes make a distinction between /dispatch/new and /acknowledgement/new. What's the reason for that?

It's how the system distinguishes between different template types when displaying the list of available templates of that type

**Comment by Polsonby** [comment]
Amended routes as suggested, thanks @mseedat-moj

File: app/controllers/cases/letters_controller.rb (hunk 3 above)
Comment (Polsonby) [reply app/controllers/cases/letters_controller.rb]: Refactored the controller - I agree about possibly using a service for this, will take a look at that in the next PR when we add Word document export

**Review by mseedat-moj — APPROVED** [review]

**Pull request merged**